app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt

```diff
@@ -0,0 +1,126 @@
+package edu.stanford.bdh.engagehf.bluetooth.data.repository
+
+import androidx.health.connect.client.records.BloodPressureRecord
+import androidx.health.connect.client.records.HeartRateRecord
+import androidx.health.connect.client.records.WeightRecord
+import ca.uhn.fhir.context.FhirContext
+import ca.uhn.fhir.context.FhirVersionEnum
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Query
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import edu.stanford.healthconnectonfhir.Loinc
+import edu.stanford.healthconnectonfhir.ObservationToRecordMapper
+import edu.stanford.spezi.core.coroutines.di.Dispatching
+import edu.stanford.spezi.core.logging.speziLogger
+import edu.stanford.spezi.module.onboarding.invitation.await
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import org.hl7.fhir.r4.model.Observation
+import javax.inject.Inject
+
+internal class ObservationRepository @Inject constructor(
+    private val firestore: FirebaseFirestore,
+    private val firebaseAuth: FirebaseAuth,
+    private val observationToRecordMapper: ObservationToRecordMapper,
+    @Dispatching.IO private val ioDispatcher: CoroutineDispatcher,
+) {
+    private val logger by speziLogger()
+
+    private val heartRateCollection =
+        ObservationCollection("heartRateObservations", Loinc.HeartRate.CODE)
+
+    private val bodyWeightObservation =
+        ObservationCollection("bodyWeightObservations", Loinc.Weight.CODE)
+
+    private val bloodPressureCollection =
+        ObservationCollection("bloodPressureObservations", Loinc.BloodPressure.CODE)
+
+    private val jsonParser by lazy {
+        FhirContext.forCached(FhirVersionEnum.R4).newJsonParser()
+    }
+
+    private val gson by lazy {
+        Gson()
+    }
+
+    suspend fun saveObservations(observations: List<Observation>) {
+        withContext(ioDispatcher) {
+            runCatching {
+                val uid = firebaseAuth.currentUser?.uid
+                    ?: throw IllegalStateException("User not authenticated")
+                val batch = firestore.batch()
+
+                observations.forEach { observation ->
+                    val collectionName = getCollectionNameForObservation(observation)
```

Comment:
Question: Should we would wrapp the content of each item here in a runCatching so that we don't miss saving the whole batch of observations in case of a single invalid one?

Reply:
Theoretically, the method could also be used to store any number of independent observations. In our case, however, it is the case that related data is always saved, where it may be more intuitive not to save the others (in our case so far a maximum of 1 more) if one fails.